FixMath2.h

```diff
@@ -0,0 +1,845 @@
+/*
+ * FixMath2.h
+ *
+ * Copyright 2023, Thomas Combriat and the Mozzi consortsium
+ *
+ * This file is part of Mozzi.
+ *
+ * Mozzi is licensed under a Creative Commons
+ * Attribution-NonCommercial-ShareAlike 4.0 International License.
+ *
+ */
+
+
+/** This file implements two fixed point number classes. These numbers can have a fractional
+part but are actually standard integers under the hood which makes calculations with them
+efficient on platforms which do not have a FPU like most micro-controllers. These numbers can be
+signed (SFixMath2) or unsigned (UFixMath2).
+
+A fixed point number has its range defined by the number of bits encoding the integer part (NI 
+in the following) and its precision by the number of bits encoding the fractional part (NF).
+
+Like standard C(++) types, the fixed point numbers defined here are following some rules:
+ - any fixed type can be converted to another *as long as the value can be represented in the destination type*. Casting to a bigger type in term of NI and NF is safe, but reducing NI can lead to an overflow if the new type cannot hold the integer value and reducing NF leads to a loss of precision.
+ - all operations between fixed point number is safe (it won't overflow) and preserve the precision. In particular:
+   - only addition, subtraction and multiplication are implemented
+   - any operation between a signed and an unsigned leads to a signed number
+   - resulting numbers will be casted to a type big enough to store the expected values. It follows that it is worth starting with types that are as small as possible to hold the initial value.
+ - all operations between a fixed point number and a native type (int, float, uint) are *not* safe. If the resulting value cannot be represented in the fixed point type it will overflow. Only addition, subtraction, multiplication and right/left shift are implemented.
+ - safe right/left shifts, which return the correct value in the correct type are implemented as .sR<shift>() and .sL<shift>() respectively, shift being the shifting amount.
+*/
```

Comment:
This should eventually contain some code examples.

If I understand correct, the advice will be to use as little a number for NI as possible, and not necessarily 8-bit aligned. For NF, I believe an 8bit-aligned number will be preferable, as this a) this does not grow, and b) shifts by 8 are more efficient on AVR. (If this is correct, it should also be stated).

Reply:
Actually, both should be as small as possible. As the integral part and the fractional part are contained inside one integer under the hood, I do not think 8bits aligned numbers are preferable for the fractional part but of course it is preferable to have the underlying int as small as possible and aligned (NI+NF = 9 forces to go 16bits under the hood). Of course if 8 shifts are more efficient on AVR that could be beneficial but such shifts occurs only at construction from a "pure" integer or float, which should not happen very often. On the long run (but I need to implant this into the relevant part of Mozzi) a few pre-defined types will be given. Note that `NF` grows with multiplication.